staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/controller.go

```diff
@@ -0,0 +1,604 @@
+func (cfgCtl *configController) OnDelete(obj interface{}) {
+	cfgCtl.triggerReload()
+}
+
+func (cfgCtl *configController) Run(stopCh <-chan struct{}) error {
+	defer cfgCtl.configQueue.ShutDown()
+	klog.Info("Starting reqmgmt config controller")
+	if ok := cache.WaitForCacheSync(stopCh, cfgCtl.plInformerSynced, cfgCtl.fsInformerSynced); !ok {
+		return fmt.Errorf("Never achieved initial sync")
+	}
+	klog.Info("Running reqmgmt config worker")
+	wait.Until(cfgCtl.runWorker, time.Second, stopCh)
+	klog.Info("Shutting down reqmgmt config worker")
+	return nil
+}
+
+func (cfgCtl *configController) runWorker() {
+	for cfgCtl.processNextWorkItem() {
+	}
+}
+
+func (cfgCtl *configController) processNextWorkItem() bool {
+	obj, shutdown := cfgCtl.configQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	func(obj interface{}) {
+		defer cfgCtl.configQueue.Done(obj)
+		if !cfgCtl.syncOne() {
+			cfgCtl.configQueue.AddRateLimited(obj)
+		} else {
+			cfgCtl.configQueue.Forget(obj)
+		}
+	}(obj)
+
+	return true
+}
+
+// syncOne attempts to sync all the config for the reqmgmt filter.  It
+// either succeeds and returns `true` or logs an error and returns
+// `false`.
+func (cfgCtl *configController) syncOne() bool {
+	all := labels.Everything()
+	newPLs, err := cfgCtl.plLister.List(all)
+	if err != nil {
+		klog.Errorf("Unable to list PriorityLevelConfiguration objects: %s", err.Error())
+		return false
+	}
+	newFSs, err := cfgCtl.fsLister.List(all)
+	if err != nil {
+		klog.Errorf("Unable to list FlowSchema objects: %s", err.Error())
+		return false
+	}
+	cfgCtl.digestConfigObjects(newPLs, newFSs)
+	return true
+}
+
+// digestConfigObjects is given all the API objects that configure
+// cfgCtl and writes its consequent new configState.  This
+// function has three loops over priority levels: one to digest the
+// given objects, one to handle old objects, and one to divide up the
+// server's total concurrency limit among the surviving priority
+// levels.
+func (cfgCtl *configController) digestConfigObjects(newPLs []*fctypesv1a1.PriorityLevelConfiguration, newFSs []*fctypesv1a1.FlowSchema) {
+	oldCfgState := cfgCtl.curState.Load().(*configState)
+	var shareSum float64 // accumulated in first two loops over priority levels
+	newCfgState := &configState{
+		priorityLevelStates: make(map[string]*priorityLevelState),
+	}
+
+	// Buffer of priority levels that need a call to QueueSet::Quiesce.
+	// See the explanation later for why these calls are held until the end.
+	newlyQuiescent := make([]*priorityLevelState, 0)
+
+	// Keep track of which mandatory objects have been digested.
+	var haveExemptPL, haveCatchAllPL, haveExemptFS, haveCatchAllFS bool
+
+	// Digest each given PriorityLevelConfiguration.
+	// Pretend broken ones do not exist.
+	for _, pl := range newPLs {
+		state := oldCfgState.priorityLevelStates[pl.Name]
+		if state == nil {
+			state = &priorityLevelState{
+				config: pl.Spec,
+			}
+		} else {
+			oState := *state
+			state = &oState
```

Comment:
looks tricky to perform shallow-copying by twiddling pointers, in favor of a new function `shallowCopyState(..)`

Reply:
shallow copying like this is pretty normal, I don't think it needs its own function.